drivers/serial/uart_gecko.c

```diff
@@ -13,23 +14,84 @@
 
 #define USART_PREFIX cmuClock_USART
 #define UART_PREFIX cmuClock_UART
-#define CLOCK_ID_PRFX2(prefix, suffix) prefix##suffix
-#define CLOCK_ID_PRFX(prefix, suffix) CLOCK_ID_PRFX2(prefix, suffix)
-#define CLOCK_USART(id) CLOCK_ID_PRFX(USART_PREFIX, id)
-#define CLOCK_UART(id) CLOCK_ID_PRFX(UART_PREFIX, id)
+#define CLOCK_USART(id) _CONCAT(USART_PREFIX, id)
+#define CLOCK_UART(id) _CONCAT(UART_PREFIX, id)
+
+/* Helper define to determine if SOC supports hardware flow control */
+#if ((_SILICON_LABS_32B_SERIES > 0) || \
+	(defined(_USART_ROUTEPEN_RTSPEN_MASK) && \
+	defined(_USART_ROUTEPEN_CTSPEN_MASK)))
+#define HW_FLOWCONTROL_IS_SUPPORTED_BY_SOC
+#endif
+
+#define DT_DRV_COMPAT silabs_gecko_uart
+
+#if (DT_INST_PROP(0, hw_flow_control) || \
+	DT_INST_PROP(1, hw_flow_control))
+#define UART_GECKO_UART_HW_FLOW_CONTROL_ENABLED
+#endif
 
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT silabs_gecko_usart
+
+#if (DT_INST_PROP(0, hw_flow_control) ||    \
+	DT_INST_PROP(1, hw_flow_control) || \
+	DT_INST_PROP(2, hw_flow_control) || \
+	DT_INST_PROP(3, hw_flow_control) || \
+	DT_INST_PROP(4, hw_flow_control) || \
+	DT_INST_PROP(5, hw_flow_control))
+#define UART_GECKO_USART_HW_FLOW_CONTROL_ENABLED
+#endif
+
+#if defined(UART_GECKO_USART_HW_FLOW_CONTROL_ENABLED) || \
+	defined(UART_GECKO_UART_HW_FLOW_CONTROL_ENABLED)
+#define UART_GECKO_HW_FLOW_CONTROL
+#endif
+
+/* Sanity check for hardware flow control */
+#if defined(UART_GECKO_HW_FLOW_CONTROL) && \
+	(!(defined(HW_FLOWCONTROL_IS_SUPPORTED_BY_SOC)))
+#error "Hardware flow control is activated for at least one UART/USART, \
+but not supported by this SOC"
+#endif
+
+/**
+ * @brief Maps init value for hw_flowcontrol
+ *
+ * @param fc_enabled Bool to enable hardware flow control
+ * @return Mapped value according to enum USART_HwFlowControl_TypeDef
+ */
+#ifdef HW_FLOWCONTROL_IS_SUPPORTED_BY_SOC
+#define MAP_HW_FLOWCONTROL_INIT_VAL(fc_enabled) \
+	(fc_enabled == 1 ? usartHwFlowControlCtsAndRts : usartHwFlowControlNone)
+#endif
+
+/**
+ * @brief Config struct for UART
+ */
 struct uart_gecko_config {
 	USART_TypeDef *base;
 	CMU_Clock_TypeDef clock;
 	uint32_t baud_rate;
+#ifdef UART_GECKO_HW_FLOW_CONTROL
+	USART_HwFlowControl_TypeDef hw_flowcontrol;
```

Comment:
I think the code would look cleaner if we had `bool` here. Then only on the line `usartInit.hwFlowControl = ` where we write to the usart init struct we should use the proper HAL type. The `MAP_HW_FLOWCONTROL_INIT_VAL` macro could be removed.

Reply:
done